### ISSUE TITLE
Bug on deeplab classifier model

### ISSUE THREAD

**Issue by lianqing11** [wasidennis/AdaptSegNet#4]
```
class Classifier_Module(nn.Module):
    def __init__(self, inplanes, dilation_series, padding_series, num_classes):
        super(Classifier_Module, self).__init__()
        self.conv2d_list = nn.ModuleList()
        for dilation, padding in zip(dilation_series, padding_series):
            self.conv2d_list.append(
                nn.Conv2d(inplanes, num_classes, kernel_size=3, stride=1, padding=padding, dilation=dilation, bias=True))

        for m in self.conv2d_list:
            m.weight.data.normal_(0, 0.01)

    def forward(self, x):
        out = self.conv2d_list[0](x)
        for i in range(len(self.conv2d_list) - 1):
            out += self.conv2d_list[i + 1](x)
            return out
```
Why is the  “return out " statement on the for loop? 
In my understanding, out is the sum of the pyramid layer , so it should be outside of the for loop.

**Comment by hfslyc** [comment]
Thank you for noticing. This is a bug also later on fixed by the reference repo ([link](https://github.com/speedinghzl/Pytorch-Deeplab/commit/cf1d48afe9c611c90f8cabba014697bb0aac5284#diff-880d3274218b27e84f82d63fff4186a9)).

The bug is essentially making our model have only 2 effective ASPP layer, and we don't think it will affect the performance too much. Still, we will retrain the models based on the fixed bug soon.

**Comment by lianqing11** [comment]
Thank for your correction. and Have you retrained the model, Does it have any improvement?

What's more, I have a question on how to select the final model? I also train the model as the procedure on read.me, but find only can get the best mIoU 39.8 at 55000 iterations and at the last iteration the mIoU only have 37+. I wonder if you can give me any suggestion on training the model.

Thank you for your time.

**Comment by John1231983** [comment]
Do you train from scratch or pretrain model? The code cannot provide good result without pretrain model

**Comment by lianqing11** [comment]
@John1231983 I train with the pretrain model(http://vllab.ucmerced.edu/ytsai/CVPR18/DeepLab_resnet_pretrained_init-f81d91e8.pth)

**Comment by wasidennis** [comment]
@lianqing11 Yes, the result may vary during the training process. We have trained the model several times and find it usually gives the IoU around 40~42 after 80k to 120k. I would suggest you to train the model longer and use a validation set to find a good model.

**Comment by dongzhuoyao** [comment]
hi, this is my result for training 80k.

|arch|mIoU|your paper|
|----|----|----|
GTA5-to-Cityscapes model (single-level,resnet101)|39.8|--|
GTA5-to-Cityscapes model (mutli-level),resnet101|41.5|42.4|

could you tell me what your training iterations  by getting your  **42.4** mIoU ?  @wasidennis

**Comment by John1231983** [comment]
One thing I want to ask, Do you think if the model obtain good result in val, then it will work well in test set? I think No, because val set is small set than testing set, hence your way may not good. We have to make a stable version. I guess the performance is not stable due to using GAN and ADAM

**Comment by wasidennis** [comment]
@John1231983 It is a good point. Of course there might be some gaps between val and test, and people now usually just use the test set to find the best model. In practice, if there is no ground truth in the test set, using the val set is one solution. This is also an interesting research direction in the future :)

**Comment by dongzhuoyao** [comment]
what's the difference in training setting between synthia and GTA?

because synthia image number(9400) is smaller than GTA(24967).

the iteration is also 8k~12K?

**Comment by wasidennis** [comment]
@dongzhuoyao Yes, the iteration number would be smaller using synthia. We did not really tune the iteration on synthia, and we found that it generally works well after 60k (only trained for 90k in total). Btw, the problem of training too long is that it will overfit more to the source data, which is not what we want.

**Comment by lianqing11** [comment]
@wasidennis  Thanks for your reply. now, I am able to reproduce the mIoU on the paper using the code.

**Comment by John1231983** [comment]
@lianqing11: any tips to reproduce it? Thanks

**Comment by lianqing11** [comment]
@John1231983  As you said, the model is not stable, so I just eval the model more frequently, such as eval it every 2000 iteration, but this is not a good solution and I also think eval on test set is needed.

**Comment by John1231983** [comment]
I see. For stable. Let try SGD in both S and D networks. Adam is just faster coverage but it is not stable. I do not like Adam
